### ISSUE TITLE
[BUG] RangeTool allows target plot to pan beyond explicit range bounds

### ISSUE THREAD

**Issue by droumis** [bokeh/bokeh#13315]
### Software versions

Python version      :  3.9.16 | packaged by conda-forge | (main, Feb  1 2023, 21:38:11) 
IPython version     :  8.14.0
Tornado version     :  6.3.2
Bokeh version       :  3.2.1
BokehJS static path :  /Users/droumis/opt/miniconda3/envs/neuro-eeg-viewer/lib/python3.9/site-packages/bokeh/server/static
node.js version     :  v20.4.0
npm version         :  9.7.2
Operating system    :  macOS-13.4.1-arm64-arm-64bit

### Jupyter notebook / Jupyter Lab version

jupyterlab          :  4.0.3

### Expected behavior

When explicit ranges are set on the target plot, the rangetool from a linked source plot should match the behavior of panning from within the target plot and not allow for panning beyond the range bounds.

### Observed behavior

The rangetool currently allows for panning beyond the explicit range bounds of the target plot.

### Example code

```Python
import numpy as np

from bokeh.layouts import column
from bokeh.models import ColumnDataSource, RangeTool
from bokeh.plotting import figure, show, output_notebook

output_notebook()

x_values = np.linspace(0, 3000, 3000)
y_values = np.cumsum(np.random.randn(3000)) + 50
source = ColumnDataSource(data=dict(x=x_values, y=y_values))

# Get the bounds for x and y axes
min_x = x_values.min()
max_x = x_values.max()
min_y = y_values.min()
max_y = y_values.max()

p = figure(height=300, width=800,
           x_range=(x_values[1500], x_values[2500]),
           y_range=(min_y, max_y),
           x_axis_label='X',
           y_axis_label='Y')

# Set the bounds for x and y axes so we cannot pan/zoom outside the data limits
p.x_range.bounds = (min_x, max_x)
p.y_range.bounds = (min_y, max_y)

p.line('x', 'y', source=source)

select = figure(height=130, width=800, y_range=p.y_range,
                x_range=(min_x, max_x),
                x_axis_label='X',
                y_axis_type=None,
                tools="", toolbar_location=None)

range_tool = RangeTool(x_range=p.x_range)
range_tool.overlay.fill_color = "navy"
range_tool.overlay.fill_alpha = 0.2

select.line('x', 'y', source=source)
select.ygrid.grid_line_color = None
select.add_tools(range_tool)

show(column(p, select))

```

### Screenshots

https://github.com/bokeh/bokeh/assets/6613202/e4ea9a22-bc38-4f40-92e8-cffdd3cf3a63

**Comment by bryevdv** [comment]
The `PanTool` defers updates to `PlotView.update_range`, which is actually what does bounds comparisons:

https://github.com/bokeh/bokeh/blob/99310e6e9140cb4f8e0655eaebe6bab156a94d2e/bokehjs/src/lib/models/tools/gestures/pan_tool.ts#L117

The `RangeTool` currently just directly sets the range `start` and `end` directly, which bypasses any bounds checking.

Probably the best solution would factor out the meat of `PlotView.update_range` into an update method on ranges themselves that checks against bounds, and then have both the `RangeTool` and `PlotView.update_range` call that. 

@droumis is this something you would be interested to look at yourself?

**Comment by bryevdv** [comment]
Well, though actually `update_range` itself defers to a `RangeManager` class that was added awhile back, but that class seems fairly coupled to "being part of a plot" so it might be some work to factor out common parts or reduce coupling. 

The "quick fix" would just be to check bounds here:

https://github.com/bokeh/bokeh/blob/99310e6e9140cb4f8e0655eaebe6bab156a94d2e/bokehjs/src/lib/models/tools/gestures/range_tool.ts#L123-L129

but obviously it's not ideal that every tool or feature has to remember to do their own bounds checks

**Comment by droumis** [comment]
@bryevdv thanks for the guidance.. I took a crack at the "quick fix" in #13321 but I don't really know what I'm doing and I haven't successfully tested it yet. I'm happy for someone else to bring it across the finish line if they want.

**Comment by mattpap** [comment]
I would say that the real issue here is that `RangeTool`'s overlay can get past frame's bounding box, which is a regression after PR #12074 (editable annotations). Previously the overlay stopped at the frame, which not only resolved the bounds issue, but also was better from UI perspective, because if the "range" plot is not pannable, then one can lose the overlay if dragging completely out of frame.
